File: Minimum_cost_spanning_tree/Prim's_algorithm/prim.py
def prim(adj_list, tw):
    
    distance = [tw for i in range(len(adj_list))]
    
    neighbour = [-1 for i in range(len(adj_list))]
    
    visited = [False for i in range(len(adj_list))]
   
    for neighbour_v, weight in adj_list[0].items():
    
        distance[neighbour_v] = weight
        neighbour[neighbour_v] = 0
    
    visited[0] = True    
        
    
    for i in range(1, len(adj_list)):
    
        minimum_weight = tw
        
        
        for j in range(1, len(distance)):
        
            if ((not visited[j]) and (distance[j] < minimum_weight)):
            
                minimum_weight = distance[j]
                
                mw_vertex = j
                
        visited[mw_vertex] = True
        
        for neighbour_v, weight in adj_list[mw_vertex].items():
        
            if (not visited[neighbour_v]) and (distance[neighbour_v] > weight):
                
                distance[neighbour_v] = weight
                neighbour[neighbour_v] = mw_vertex             
    
    for i in range(1, len(neighbour)):
        
        print(neighbour[i], " <--> ", i, " with distance ", distance[i])
        
                 
def check_vertex(vertex, nov):
    
    while(1):
        
        if (vertex >= nov or vertex < 0):
            
            print("Invalid vertex: ", vertex)
            
            vertex = int(input("Re-enter the vertex: "))
        
        else:
            return vertex


def read_edges(nov, noe, adj_list):                  # noe = no_of_edges
    
    tw = 1
    
    print("\nProvide the edges and their weights: \nvertex1 vertex2 weight", end="\n")
    
    while (noe):

        vertex = list(input().split(" "))
        source, destination, weight = int(vertex[0]), int(vertex[1]), int(vertex[2])     
    
        source = check_vertex(source, nov)
        destination = check_vertex(destination, nov)
        
        tw += abs(weight)
        
        adj_list[source][destination] = weight
        adj_list[destination][source] = weight
    
        
        noe -= 1
    
    return (tw)         

###################################################

no_of_vertices = int(input("Enter the no of vertices: "))

adj_list = [{} for i in range(0, (no_of_vertices))]

no_of_edges = int(input("\nEnter the no of edges: "))    

total_weight = read_edges(no_of_vertices, no_of_edges, adj_list)

print("\n\nCalling Prim's algorithm...", end="\n")

prim(adj_list, total_weight)          # Calling the Prim's algorithm        

File: Minimum_cost_spanning_tree/Prim's_algorithm/prim.cpp
#include <iostream>
#include <vector>
#include <map>


void prim(std::vector<std::map<int,int>>& adj_list, int& twc)
{

	std::vector<int> distance(adj_list.size(), twc);

	std::vector<bool> visited(adj_list.size(), false);

	std::vector<int> neighbour(adj_list.size(), -1);

	for (auto neighbour_of_vertex : adj_list[0])
	{

		distance[neighbour_of_vertex.first] = neighbour_of_vertex.second; 
		neighbour[neighbour_of_vertex.first] = 0;

	}

	visited[0] = true;
	distance[0] = 0;
	neighbour[0] = 0;

	for (int i = 1; i < adj_list.size(); i++)
	{

		int minimum_weight = twc;
		int selected_vertex;

		for (int vn = 1; vn < adj_list.size(); vn++) 
		{

			if (!visited[vn] && distance[vn] < minimum_weight)  
			{

				minimum_weight = distance[vn];
				selected_vertex = vn;

			}
		}

		visited[selected_vertex] = true;

		for (auto neighbour_of_vertex : adj_list[selected_vertex])
		{

			if (!visited[neighbour_of_vertex.first] and (distance[neighbour_of_vertex.first] > neighbour_of_vertex.second)) 
			{

				neighbour[neighbour_of_vertex.first] = selected_vertex;
				distance[neighbour_of_vertex.first] = neighbour_of_vertex.second;

			}
		}

	} 

	std::cout << "\nThe following edges and their distances form the minimum cost tree" << std::endl;

	for (int i = 1; i < neighbour.size(); i++)
	{

		std::cout << neighbour[i] << " <--> " << i << " with distance = " << distance[i] << std::endl;

	}

}


void check_vertex(int& v, int& vc)	// check_vertex starts
{
	while (v >= vc || v < 0)
	{
		std::cout << "Invalid vertex no: " << v << std::endl;
		std::cout << "Re enter the vertex: " << std::endl;
		std::cin >> v;
	}
}	// check_vertex ends


void read_vertex_and_edge_count(int& v_count, int& e_count)
{

	std::cout << "Enter the no of vertices:" << std::endl;
	std::cin >> v_count;
	std::cout << "Enter the no of edges: " << std::endl;
	std::cin >> e_count;

}


void read_the_edges(std::vector<std::map<int,int>>& adj, int& ec, int& v_count, int& twc)
{

	int source;
	int destination;
	int weight;
	std::cout << "\nEnter the edges and their weights: " << std::endl;

	while (ec--)
	{
		std::cin >> source >> destination >> weight;

		check_vertex(source, v_count);
		check_vertex(destination, v_count);

		weight < 0 ? twc += ((-1) * (weight)) : twc += weight;         // Representing the concept of infinity

		adj[source].insert({destination, weight});

		adj[destination].insert({source, weight});

	}

}


int main()
{

	int vertex;
	int vertices_count = 0;
	int edges_count = 0;
	int total_weight_count = 1;

	read_vertex_and_edge_count(vertices_count, edges_count);

	std::vector<std::map<int,int>> adjacency_list(vertices_count);
	read_the_edges(adjacency_list, edges_count, vertices_count, total_weight_count);

	prim(adjacency_list, total_weight_count);             // CALLING THE PRIM'S ALGORITHM

	return 0;
}
